diff --git a/hack/make/.binary b/hack/make/.binary
--- a/hack/make/.binary
+++ b/hack/make/.binary
@@ -74,7 +74,7 @@
 	if [[ " $BUILDFLAGS " != *" -race "* ]]; then
 		# -buildmode=pie is not supported on Windows and Linux on mips and riscv64.
 		case "$(go env GOOS)/$(go env GOARCH)" in
-			windows/* | linux/mips* | linux/riscv*) ;;
+			windows/* | linux/mips* | linux/riscv* | linux/ppc64) ;;
 
 			*)
 				BUILDFLAGS+=("-buildmode=pie")
