From 26c53d690eee10fbe9ee948165547388ccd42d57 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Micha=C5=82=20G=C3=B3rny?= <mgorny@gentoo.org>
Date: Wed, 3 Feb 2021 10:14:49 +0100
Subject: [PATCH 04/12] ssl: Hard-disable SSLv3 to avoid automagic deps

Bug: https://bugs.gentoo.org/767886
---
 Modules/_ssl.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/Modules/_ssl.c b/Modules/_ssl.c
index 6071202104..d11ec05e6d 100644
--- a/Modules/_ssl.c
+++ b/Modules/_ssl.c
@@ -65,6 +65,10 @@
 #  error "OPENSSL_THREADS is not defined, Python requires thread-safe OpenSSL"
 #endif
 
+#ifndef OPENSSL_NO_SSL3
+#  define OPENSSL_NO_SSL3 1
+#endif
+
 
 
 struct py_ssl_error_code {
-- 
2.38.1

